### PR TITLE
Allow menus to be dynamically positioned.

### LINKED ISSUE #2517: Call Options menu doesn't dynamically scale/position when the window size is changed
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!

**Upwork job link** - https://www.upwork.com/jobs/~01ba0dff1eec1e5790
___

## Expected Result:
When you resize the Expensify Cash window, the Call Options menu should dynamically scale within the window.

## Actual Result:
When you resize the Expensify Cash application, the Call Options menu does not properly scale with the window. Sometimes it gets cut off, the position is off, and the spacing is wrong. 

## Action Performed:
1. Open Expensify.cash on Web or Desktop
2. Log into your account
3. Click into a chat with another user 
4. Click the 📞 phone icon in the top-right, and see the Call Option menu
5. Click on any edge of the window and change the size of the window. See that the Call Option menu isn't dynamically moving with the size of the window.

Yes, Expensify is still usable and you can size the window in a way that allows you to select the options.

## Platform:
Where is this issue occurring?

 - [X] Web
 - [X] Desktop App
- [ ] iOS
- [ ] Android
- [ ] Mobile Web

**Version Number:** tested most recently on 1.0.22-1 (1.0.22-1)
**Notes/Photos/Videos:** See GIF recording below:
dabuggy

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/159934

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

### PR DESCRIPTION
**Pull request by parasharrajat**
Please review @roryabraham

### Details
1. Added a new prop `onDimensionsChange` to `WithWindowDimensions` which allows attaching change Listener.
2. Debounced the Dimensions change event on `WithWindowDimensions` by 200ms.
2. Using the above, set the recalculation of the Call menu and Emoji Picker.

### Fixed Issues

Fixes #2517

### Tests / QA Steps
1. Open Expensify.cash on Web or Desktop
2. Log into your account
3. Click into a chat with another user
4. Click the telephone_receiver phone icon in the top-right, and see the Call Option menu
5. Click on any edge of the window and change the size of the window. See that the Call Options menu dynamically moving with the size of the window.
4. Click the Emoji picker icon on the report composer.
5. Click on any edge of the window and change the size of the window. See that the Emoji picker menu dynamically moving with the size of the window.

### Tested On

- [x] Web
- [ ] Mobile Web
- [x] Desktop
- [ ] iOS
- [ ] Android

#### Web / Desktop

https://user-images.githubusercontent.com/24370807/117215199-872aa100-ae1b-11eb-8af1-54581936bc30.mp4

### CODE REVIEW

**Comment by roryabraham** [comment]
Hmmm @parasharrajat I thought we discussed in [the linked issue](https://github.com/Expensify/Expensify.cash/issues/2517) that we were going to move the dimensions informations and listeners into Onyx, and deprecate the `withWindowDimensions` HOC?

**Comment by parasharrajat** [comment]
Sure. Will do.

**Comment by parasharrajat** [comment]
@roryabraham I have a suggestion. Can we use the React Context for WindowDimensions? https://reactjs.org/docs/context.html.

I still have doubts. I just want to achieve the best possible thing here. I understand Onyx is the way to go currently in the app.

1. What will happen when storage is full and Dimensions are only a runtime thing and having a read-write headache from Onyx ix worth investing in just because of simplicity.

**Comment by roryabraham** [comment]
So my take on this is that we definitely don't want to use React Context. It's a pattern we haven't used anywhere else the codebase and not one that I think we want to encourage. I'm sure I have some biases, but I think they're just sort of complex and not that commonly used 🤷🏼‍♂️

> What will happen when storage is full and Dimensions are only a runtime thing and Is having a read-write headache from Onyx worth investing in just because of simplicity?

To quote @tgolen, "this is just a rumor of a problem for now" 😄 Expensify.cash is built on Onyx, so I think the way that we should approach this is not to avoid using it because we think it might have some potential pitfalls. Instead we should use it, and then when there are problems, adjust it and make it work for our needs. It would be easy to adjust Onyx down the line to have some values only be stored in-memory 🙂

**Comment by parasharrajat** [comment]
Thanks, @roryabraham. Let me do it that way.

**Comment by parasharrajat** [comment]
@roryabraham Updated.
Do I need to refactor all the `withWindowDimensions` instances with `withOnyx` HOC, then How should I manage the direct props.
```
 windowHeight: window.height,
 windowWidth: window.width,
 isSmallScreenWidth,
```
instead of currently what I have done
```
dimensions: {
 	windowHeight: window.height,
 	windowWidth: window.width,
 	isSmallScreenWidth,
 }
```

**Comment by parasharrajat** [comment]
@roryabraham Any thoughts?

**Comment by roryabraham** [comment]
Hey @parasharrajat, I started reviewing this but am talking it over with @marcaaron currently and we're going to discuss some more in the linked issue.

**Comment by parasharrajat** [comment]
@roryabraham After Marc's comments what should I go with. I personally think that Dimensions are runtime thing and we should not involve Onyx as Onyx go through persistent storage.
Hopefully we come to a decision and I can quickly wrap up this issue.

**Comment by parasharrajat** [comment]
I found the following issues.
1. Messages that were repeated were on the same tab. (The first solution fix this part)
2. I test messages from multiple tabs. But it was not covered so I pushed a new implementation that covers this as well. I think this one is simpler ad more clear.

What the code is doing?

### OLD Architecture

####Single Tab(Client) case

When a user is offline and we get requests, We take retry requests from the normal queue and push them to Persisted Queue(We have to). We keep the copy of those requests in the normal queue as well. Thus when we get back online. requests will be sent from Persisted Queue & normal queue.(duplicates) 

#### Multi Client Case

We get requests while offline, **Single Tab(Client) case** will be repeated for all tabs. But Persisted Queue is shared among tabs so we have multi-tab duplicated requests now. With respect to this, due to when we push to Persisted queue. Onyx.connect will be called and it will push the requests to each tabs normal queue for the first time then `didLoadPersistedRequests` will block further calls.

So Recent changes change the architecture in the following way

### Proposed Architecture

1. We receive requests when the user is offline, we push retry requests to persisted queue and remove them from the normal queue. Now `Onyx.connect` will be called for all tabs but `isOffline` check will block that. For all tabs same behaviour will happen. and Persisted queue has single instance of all the requests from all tabs.

2. Now the internet is back, `Onyx.connect` is called for all tabs but only the leader client will push those requests in the `networkqueue` thus no duplicates.

**OR**

3. If the user closed the tab, we still have those requests on Persisted Queue, Onyx.connect will be called for all tabs, Again the only leader will process them. 

I think it is a simpler approach and more cleaner but I could be missing something. Otherwise, I don't think fixing the linked issue considering single tab behavior is good it would be temporary. We can wait until we are looking at the whole scenerio.

**Comment by marcaaron** [comment]
@parasharrajat I think this comment was meant for another issue/PR

**Comment by roryabraham** [comment]
@parasharrajat yep I agree with you and @marcaaron here. Sorry for the roundabout approach – I got caught up with a problem that wasn't real which led to an over-engineered solution. So yeah let's go for @marcaaron's simple solution he posted in the issue.

File: src/pages/home/report/ReportActionCompose.js [hunk 1]
```diff
@@ -227,17 +238,24 @@ class ReportActionCompose extends React.Component {
...
-     * @param {Object} [event] - A press event.
      */
-    showEmojiPicker(event) {
+    showEmojiPicker() {
         this.textInput.blur();
-        this.state.emojiPopoverAnchorPosition = {
-            horizontal: event.nativeEvent.pageX,
-            vertical: event.nativeEvent.pageY,
-        };
         this.setState({isEmojiPickerVisible: true});
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:244]: Here we were showing the Picker based on mouse position. As a result, Piker will be shown on different locations based on click position. Let me know if we want this behavior. I removed it as other Menus does not have this behavior keeping consistency.

**Comment by parasharrajat** [comment]
Updated Thanks.

**Review by roryabraham — APPROVED** [review]
Okay, this looks good 👍 

I also did some basic regression testing of the menus on iOS and mobile safari and didn't see any errors or issues, so we should be good.

**Pull request merged**